mirgecom/navierstokes.py

```diff
@@ -0,0 +1,416 @@
+
+    if grad_cv is None:
+        grad_cv = grad_cv_operator(
+            discr, gas_model, boundaries, state, time=time,
+            numerical_flux_func=gradient_numerical_flux_func,
+            quadrature_tag=quadrature_tag,
+            operator_states_quad=operator_states_quad)
+
+    # Communicate grad(CV) and put it on the quadrature domain
+    # FIXME/ReviewQuestion: communicate grad_cv - already on quadrature dom?
+    grad_cv_interior_pairs = [
+        # Get the interior trace pairs onto the surface quadrature
+        # discretization (if any)
+        _interp_to_surf_quad(tpair)
+        for tpair in interior_trace_pairs(discr, grad_cv)
+    ]
+
+    # }}} Compute grad(CV)
+
+    # {{{ === Compute grad(temperature) ===
+
+    if grad_t is None:
+        grad_t = grad_t_operator(
+            discr, gas_model, boundaries, state, time=time,
+            numerical_flux_func=gradient_numerical_flux_func,
+            quadrature_tag=quadrature_tag,
+            operator_states_quad=operator_states_quad)
+
+    # Create the interior face trace pairs, perform MPI exchange, interp to quad
+    grad_t_interior_pairs = [
+        # Get the interior trace pairs onto the surface quadrature
+        # discretization (if any)
+        _interp_to_surf_quad(tpair)
+        for tpair in interior_trace_pairs(discr, grad_t)
+    ]
+
+    # }}} compute grad(temperature)
+
+    # {{{ === Navier-Stokes RHS ===
+
+    # Compute the volume term for the divergence operator
+    vol_term = (
+
+        # Compute the volume contribution of the viscous flux terms
+        # using field values on the quadrature grid
+        viscous_flux(state=vol_state_quad,
+                     # Interpolate gradients to the quadrature grid
+                     grad_cv=op.project(discr, dd_base, dd_vol_quad, grad_cv),
+                     grad_t=op.project(discr, dd_base, dd_vol_quad, grad_t))
+
+        # Compute the volume contribution of the inviscid flux terms
+        # using field values on the quadrature grid
+        - inviscid_flux(state=vol_state_quad)
+    )
+
+    # Compute the boundary terms for the divergence operator
+    bnd_term = (
+
+        # All surface contributions from the viscous fluxes
+        viscous_boundary_flux_for_divergence_operator(
+            discr, gas_model, boundaries, inter_elem_bnd_states_quad,
+            domain_bnd_states_quad, grad_cv, grad_cv_interior_pairs,
+            grad_t, grad_t_interior_pairs, quadrature_tag=quadrature_tag,
+            numerical_flux_func=viscous_numerical_flux_func, time=time)
+
+        # All surface contributions from the inviscid fluxes
+        - inviscid_boundary_flux_for_divergence_operator(
+            discr, gas_model, boundaries, inter_elem_bnd_states_quad,
+            domain_bnd_states_quad, quadrature_tag=quadrature_tag,
+            numerical_flux_func=inviscid_numerical_flux_func, time=time)
+
+    )
+
+    return div_operator(discr, dd_vol_quad, dd_faces_quad, vol_term, bnd_term)
```

Comment:
I still find that the use of `div_operator` here obscures what's really going on. But that's not for this review, I guess.

Reply:
(https://github.com/illinois-ceesd/mirgecom/issues/598)